src/main/java/androidx/text/SpannableString.kt

```diff
@@ -70,3 +70,59 @@ inline operator fun Spannable.set(range: IntRange, span: Any) {
     // This looks weird, but endInclusive is just the exact upper value.
     setSpan(span, range.start, range.endInclusive, SPAN_INCLUSIVE_EXCLUSIVE)
 }
+
+/**
+ * Apply a [span] to the [substring] in the text.
+ *
+ * ```
+ * val s = "Hello, World!".toSpannable()
+ * s.overlay("Hello", UnderlineSpan())
+ * ```
+ *
+ * Note: Does nothing if the substring is not found.
+ *
+ * @param ignoreCase true to ignore character case when matching a string. By default false.
+ * @param startIndex starting index to search the string. Default: 0
+ *
+ * @return the end index (exclusive) of the applied span, or -1 if not found.
+ *
+ * @see Spannable.setSpan
+ */
+fun Spannable.overlay(
+    substring: String,
+    span: Any,
+    ignoreCase: Boolean = false,
+    startIndex: Int = 0
+): Int {
+    val start = indexOf(substring, startIndex, ignoreCase)
+    if (start == -1) return start
+    val end = start + substring.length
+    setSpan(span, start, end, Spannable.SPAN_INCLUSIVE_INCLUSIVE)
+    return end
+}
+
+/**
+ * Apply a [span] to all occurrences of [substring] in the text.
+ *
+ * ```
+ * val s = "Hello, World! I'm World".toSpannable()
+ * s.overlayAll("World", UnderlineSpan())
+ * ```
+ *
+ * Note: Does nothing if the substring is not found.
+ *
+ * @param ignoreCase true to ignore character case when matching a string. By default false.
+ * @param startIndex starting index to search the string. Default: 0
+ *
+ * @see Spannable.setSpan
+ */
+inline fun Spannable.overlayAll(
+    substring: String,
+    ignoreCase: Boolean = false,
+    startIndex: Int = 0,
+    crossinline spanBuilder: () -> Any
+) {
+    generateSequence(startIndex) { overlay(substring, spanBuilder(), ignoreCase, it) }
+        .takeWhile { length >= it || it != -1 }
+        .toList()
```

Comment:
A sequence seems overkill. You suffer from boxing on all the integers, allocation of the intermediate operator wrappers, and allocation of the list which is thrown away.

Let's do a simple `while` loop instead?

Reply:
All good things to learn, thanks! Using a simple while loop now, and also fixed the halting problem with using `||` instead of `&&` for multiple negative conditions 🤦‍♂️ .